### PR TITLE
Add signature help to Typescript server

### PR DESCRIPTION
**Pull request by dbaeumer**
Enclosed a pull request to add signature help support to the TypeScript server. I tried to follow the style I found in the existing code (e.g. expose the language service API, no extra abstraction).

There is one thing I left open: I don't know the meaning of applicableSpan in SignatureHelpItems. Can someone of you fill in the comment.

Thanks Dirk

### CODE REVIEW

File: src/server/session.ts [hunk 1]
```diff
@@ -577,6 +578,35 @@ module ts.server {
...
+            
+            var compilerService = project.compilerService;
+            var position = compilerService.host.lineOffsetToPosition(file, line, offset);
+            var helpItems = compilerService.languageService.getSignatureHelpItems(file, position);
+            if (!helpItems) {
+                return undefined;
+            }
+            
+            var span = helpItems.applicableSpan;
+            var result:protocol.SignatureHelpItems = {
```
Comment (DanielRosenwasser) [inline src/server/session.ts]: Add a space after the colon

**Comment by DanielRosenwasser** [comment]
Hey @dbaeumer, `applicableSpan` is simply the span for which signature help should appear on a signature - I believe the idea is that if you request signature help and move your cursor to another location within that span, you should continue seeing the same signature help dialog.

For details on applicable spans, see the [applicable span calculation for call expressions](
https://github.com/Microsoft/TypeScript/blob/35040b9a8514df9cc209398de29fa0757a38541f/src/services/signatureHelp.ts#L396) and [for tagged templates](https://github.com/Microsoft/TypeScript/blob/35040b9a8514df9cc209398de29fa0757a38541f/src/services/signatureHelp.ts#L410).

File: src/server/session.ts [hunk 2]
```diff
@@ -577,6 +578,35 @@ module ts.server {
...
+                items: helpItems.items,
+                applicableSpan: {
+                    start: compilerService.host.positionToLineOffset(file, span.start),
+                    end: compilerService.host.positionToLineOffset(file, span.start + span.length)
```
Comment (DanielRosenwasser) [inline src/server/session.ts]: You guys seem to do something similar throughout the function, you might want to consider just making a helper that returns a `{start: number, end: number}`.

**Comment by DanielRosenwasser** [comment]
Apart from the nit about the space after the colon, this looks good to me. I'll let @steveluc give the :+1: though.

**Comment by steveluc** [comment]
This is great.  Thanks.  Could you consider changing the name to SignatureHelpParameter to SignatureHelpArgs.  The rest are named with the Args suffix.

**Comment by steveluc** [comment]
Disregard previous comment.  I see the naming is consistent and SignatureHelpParameter is part of the response.

:+1:

File: src/server/session.ts [hunk 3]
```diff
@@ -790,6 +820,11 @@ module ts.server {
                                                            completionDetailsArgs.entryNames,completionDetailsArgs.file);
                         break;
                     }
+                    case CommandNames.SignatureHelp: {
```
Comment (mhegazy) [inline src/server/session.ts]: Put the comment on the wrong PR. so putting it here again :)

Can you also update client.ts and add a test for signature help in tests\cases\fourslash\server. I would copy a signature help test from tests\cases\fourslash to start.

File: src/server/protocol.d.ts [hunk 4]
```diff
@@ -593,6 +593,122 @@ declare module ts.server.protocol {
...
+        documentation: SymbolDisplayPart[];
+    }
+    
+    /**
+     * Signation help items found in the response of a signature help request.
+     */
+    export interface SignatureHelpItems {
+      
+        /**
+         * The signaure help items.
```
Comment (DanielRosenwasser) [inline src/server/protocol.d.ts]: signature

File: src/server/protocol.d.ts [hunk 5]
```diff
@@ -593,6 +593,122 @@ declare module ts.server.protocol {
...
+        parameters: SignatureHelpParameter[];
+                
+        /**
+         * The signature's documentation
+         */
+        documentation: SymbolDisplayPart[];
+    }
+    
+    /**
+     * Signation help items found in the response of a signature help request.
```
Comment (DanielRosenwasser) [inline src/server/protocol.d.ts]: Signature

**Comment by dbaeumer** [comment]
Thanks for all the comments. Will work on the test and fix the typos and provide an updated. pull request.

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**